### PR TITLE
macos: rebuild editor from the ground up

### PR DESCRIPTION
**Pull request by Parth**

### CODE REVIEW

File: libs/editor/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+target/
+.idea/
+.DS_Store
+SwiftEditor/Libs/
```
Comment (steverusso) [inline libs/editor/.gitignore:4]: Let's drop this entire file or move the contents to the root level gitignore.

File: libs/editor/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+In order of importance / when we would need it:
+
++ ability to do basic spell check
++ The ability to render markdown inline
++ Ability to automatically insert characters (auto-insert bullets)
++ The ability to render known characters as custom glyphs (`+` becomes `•`)
++ Ability to render images / drawings inline
++ The ability to render check lists for todo lists
++ Checklists be interact-able
++ Ability to be able to create multiple cursors that represent multiple concurrent collaborators
```
Comment (steverusso) [inline libs/editor/README.md:20]: If this file is staying: consistency in first letter capitalization.

File: libs/editor/egui_editor/include/egui_editor.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+typedef struct ThemedColor ThemedColor;
+
+
+
+
+
```
Comment (steverusso) [inline libs/editor/egui_editor/include/egui_editor.h:11]: Is this really generated?

File: libs/editor/egui_editor/rustfmt.toml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+use_small_heuristics = "Max"
+chain_width = 60
+fn_args_layout = "Compressed"
+use_field_init_shorthand = true
```
Comment (steverusso) [inline libs/editor/egui_editor/rustfmt.toml:4]: Also a file I'd say to drop or merge with root level one.

File: libs/editor/egui_editor/src/apple/apple_ffi.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                limits: adapter.limits(),
+            },
+            None,
+        )
+        .await;
+    match res {
+        Err(err) => {
+            panic!("request_device failed: {:?}", err);
+        }
+        Ok(tuple) => (adapter, tuple.0, tuple.1),
```
Comment (steverusso) [inline libs/editor/egui_editor/src/apple/apple_ffi.rs:115]: Would be clearer to name the tuple's values

File: libs/editor/egui_editor/src/ast.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
+use crate::cursor_types::DocByteOffset;
+use crate::element::Element;
+use pulldown_cmark::{Event, OffsetIter, Options, Parser};
+use std::ops::Range;
+
+#[derive(Default, Debug)]
+pub struct AST {
```
Comment (steverusso) [inline libs/editor/egui_editor/src/ast.rs:7]: `Ast` (this is Rust).

File: libs/editor/egui_editor/include/egui_editor.h (hunk 3 above)
Comment (Parth) [reply libs/editor/egui_editor/include/egui_editor.h:11]: lol it is, but it shouldn't be committed

File: libs/editor/egui_editor/rustfmt.toml (hunk 4 above)
Comment (Parth) [reply libs/editor/egui_editor/rustfmt.toml:4]: yep dropping

File: libs/editor/egui_editor/src/cursor_types.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        self.0.partial_cmp(other)
+    }
+}
+
+impl Add<usize> for DocCharOffset {
+    type Output = DocCharOffset;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        let sum = self.0 + rhs;
+        DocCharOffset(sum)
```
Comment (steverusso) [inline libs/editor/egui_editor/src/cursor_types.rs:85]: `Self`, and there are a few more cases below

File: libs/editor/egui_editor/src/theme.rs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    pub fn link(&self) -> Color32 {
+        self.link.unwrap_or(BLUE).get(self.current_theme)
+    }
+
+    pub fn update(&mut self, ui: &egui::Ui) {
+        if ui.visuals().dark_mode {
+            self.current_theme = Theme::Dark
+        } else {
+            self.current_theme = Theme::Light
+        }
```
Comment (steverusso) [inline libs/editor/egui_editor/src/theme.rs:114]: `self.current_theme = if { .. } else { .. };` would be a bit more concise

File: libs/editor/egui_editor/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        };
+        self.surface.configure(&self.device, &surface_config);
+    }
+}
+
+// standup
+// + [x] made scrolling nicer
+// + [x] cleaned up a bit
+// + [x] added types for the various cursors going on
+// + [ ] up down navigation
```
Comment (steverusso) [inline libs/editor/egui_editor/src/lib.rs:130]: If only there were a note taking app where you could put this

File: libs/editor/egui_editor/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+impl WgpuEditor {
+    pub fn frame(&mut self) {
+        self.configure_surface();
+        let output_frame = match self.surface.get_current_texture() {
+            Ok(frame) => frame,
+            Err(wgpu::SurfaceError::Outdated) => {
+                // This error occurs when the app is minimized on Windows.
+                // Silently return here to prevent spamming the console with:
+                // "The underlying surface has changed, and therefore the swap chain must be updated"
+                panic!()
```
Comment (steverusso) [inline libs/editor/egui_editor/src/lib.rs:47]: I would still include some kind of message

File: libs/editor/egui_editor/src/theme.rs (hunk 8 above)
Comment (Parth) [reply libs/editor/egui_editor/src/theme.rs]: bb24b906

File: libs/editor/egui_editor/src/ast.rs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        let new_child_idx = self.nodes.len();
+        self.nodes.push(node);
+        self.nodes[parent_idx].children.push(new_child_idx);
+        new_child_idx
+    }
+
+    fn look_back_whitespace(s: &str, index: DocByteOffset) -> usize {
+        if index == 0 {
+            return 0;
+        };
```
[the remark this answers is not in the document]
Comment (Parth) [reply libs/editor/egui_editor/src/ast.rs]: bb24b906

File: libs/editor/egui_editor/src/ast.rs [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+use std::ops::Range;
+
+#[derive(Default, Debug)]
+pub struct Ast {
+    pub nodes: Vec<ASTNode>,
+    pub root: usize,
+}
+
+#[derive(Default, Debug)]
+pub struct ASTNode {
```
[the remark this answers is not in the document]
Comment (Parth) [reply libs/editor/egui_editor/src/ast.rs]: bb24b906

File: libs/editor/egui_editor/src/apple/apple_ffi.rs (hunk 5 above)
Comment (tvanderstad) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs:121]: yeah bit of a nit but would like to see more destructuring assignment from you in general @Parth; aside from getting the `usize` out of our 1-tuple structs like `DocCharOffset`, I view every instance of `.0` or `.1` like this as a code smell

Comment (Parth) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs]: yeah this bit was copied and pasted from samples elsewhere, but agreed

File: libs/editor/egui_editor/src/apple/apple_ffi.rs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        rpass,
+        screen,
+        context,
+        raw_input: Default::default(),
+        editor,
+    };
+
+    obj.frame();
+
+    // TODO we need to free this memory
```
Comment (steverusso) [inline libs/editor/egui_editor/src/apple/apple_ffi.rs:47]: Has this been taken care of yet?

File: libs/editor/egui_editor/src/apple/apple_ffi.rs [hunk 14]
```diff
@@ -0,0 +1,123 @@
+use crate::apple::keyboard::NSKeys;
+use crate::{Editor, WgpuEditor};
+use egui::{Context, Event};
+use egui_wgpu_backend::{wgpu, ScreenDescriptor};
+use std::ffi::{c_char, c_void, CStr};
+
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn init_editor(
+    metal_layer: *mut c_void, //, width: u32, height: u32, scaling: u32,
```
Comment (steverusso) [inline libs/editor/egui_editor/src/apple/apple_ffi.rs:10]: Can this comment go?

File: libs/editor/egui_editor/src/ast.rs [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+                    self.push_child(current_idx, AstNode::new(Element::InlineCode, range));
+                }
+                Event::End(done) => {
+                    if let Some(done) = Element::from_tag(done) {
+                        if done == self.nodes[current_idx].element {
+                            break;
+                        }
+                    }
+                }
+                _ => {} // todo: there are some interesting events ignored (rules, tables, etc)
```
Comment (steverusso) [inline libs/editor/egui_editor/src/ast.rs:65]: I would just make this a normal comment above this line and drop the 'todo' part, but just resolve if you want to leave as is.

File: libs/editor/egui_editor/src/ast.rs [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+//    other elements do (TEST_MARKDOWN_13 vs TEST_MARKDOWN_25)
+// 2. inconsistent code block behavior, a code block that is defined with spaces in front (rather
+//    than by a code fence) begins at the first character after the spaces, but the space characters
+//    are not absorbed anywhere else. And this code block includes a \n at the end unlike the code
+//    fence block
+// 3. the indentation (whitespace) at the start of an item is not part of the item
+// 4. a \n\n at the end of an item remains part of that item even if it's a \n\ntest
+//
+// These things are either going to serve as motivation for a custom editor down the road, or an
+// explanation for strange things like look_back_whitespsace
```
Comment (steverusso) [inline libs/editor/egui_editor/src/ast.rs:135]: I would move this into the readme or eng or shared lockbook folder.

File: libs/editor/egui_editor/src/galley.rs [hunk 17]
```diff
@@ -0,0 +1,160 @@
+use crate::cursor_types::{DocByteOffset, DocCharOffset};
+use crate::editor::Editor;
+use crate::layout_job::{Annotation, LayoutJobInfo};
+use egui::epaint::text::cursor::Cursor;
+use egui::{Color32, Galley, Pos2, Rect, Sense, Stroke, Ui, Vec2};
+use std::ops::Range;
+use std::sync::Arc;
+
+// todo: maybe a nice VisualAppearence struct is in order?
```
Comment (steverusso) [inline libs/editor/egui_editor/src/galley.rs:9]: Is this a stale todo since `VisualAppearence` struct is a thing in theme?

**Review by steverusso — CHANGES_REQUESTED** [review]

File: libs/editor/egui_editor/src/galley.rs (hunk 17 above)
Comment (tvanderstad) [reply libs/editor/egui_editor/src/galley.rs:9]: there is such a struct but this isn't included in it; I have changes I'm working on to address it though

File: libs/editor/egui_editor/src/apple/apple_ffi.rs (hunk 13 above)
Comment (Parth) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs:47]: no, this todo will need to stay for a bit

File: libs/editor/egui_editor/src/apple/apple_ffi.rs (hunk 14 above)
Comment (Parth) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs:10]: yes, I'll remove

File: libs/editor/egui_editor/src/ast.rs (hunk 15 above)
Comment (Parth) [reply libs/editor/egui_editor/src/ast.rs:65]: resolving

File: libs/editor/egui_editor/src/ast.rs (hunk 16 above)
Comment (Parth) [reply libs/editor/egui_editor/src/ast.rs:135]: Can you say more words? Feels like relevant docs for this file and the functions in here

File: libs/editor/egui_editor/src/apple/apple_ffi.rs (hunk 5 above)
Comment (Parth) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs]: I'll address shortly

File: libs/editor/egui_editor/src/ast.rs (hunk 16 above)
Comment (steverusso) [reply libs/editor/egui_editor/src/ast.rs:135]: I don't feel _too_ strongly about it, just that "shortcomings / issues of a certain dep / technique" that goes beyond a localized sentence or two should probably be literature that lives somewhere else besides right in the code itself.

File: libs/editor/egui_editor/src/apple/apple_ffi.rs (hunk 14 above)
Comment (Parth) [reply libs/editor/egui_editor/src/apple/apple_ffi.rs]: fixed

File: libs/editor/egui_editor/src/ast.rs (hunk 16 above)
Comment (Parth) [reply libs/editor/egui_editor/src/ast.rs:135]: for me this feels like the place for now

File: libs/editor/egui_editor/src/offset_types.rs [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 += rhs.0
+    }
+}
+
+impl SubAssign<RelByteOffset> for RelByteOffset {
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 -= rhs.0
+    }
+}
```
Comment (steverusso) [inline libs/editor/egui_editor/src/offset_types.rs:51]: newline

File: libs/editor/egui_editor/src/test_input.rs [hunk 19]
```diff
@@ -0,0 +1,179 @@
+pub static TEST_MARKDOWN: &str = TEST_MARKDOWN_42;
```
Comment (steverusso) [inline libs/editor/egui_editor/src/test_input.rs:1]: This file staying around for now?

**Pull request merged**